File: leetcode2025/2025_03/leetcode2025-03-01.cpp
#include <iostream>
#include <vector>
#include <string>
using namespace std;

class Solution {
private:
    string str;
    int n;
    vector<vector<string>> res;
    vector<string> path;

    bool check(string tmp){
        for (int i=0; i<tmp.size()/2; i++){
            if (tmp[i] != tmp[tmp.size()-i-1]){
                return false;
            }
        }
        return true;
    }

    void dfs(int lastIndx){
        if (lastIndx == n-1){
            res.push_back(path);
            return;
        }
        for (int i=lastIndx+1; i<n; i++){
            string tmp = str.substr(lastIndx+1, i-lastIndx);
            if (check(tmp)){
                path.push_back(tmp);
                dfs(i);
                path.erase(path.end()-1);
            }
        }
    }
public:
    vector<vector<string>> partition(string s) {
        str = s;
        n = s.size();
        dfs(-1);
        return res;
    }
};

int main(){
    Solution sl;
    string s = "aab";
    vector<vector<string>> res = sl.partition(s);
    for (auto& i : res){
        for (auto& j : i){
            cout << j << " ";
        }
        cout << endl;
    }
}

File: leetcode2025/2025_03/leetcode2025-03-01.py
from typing import List

class Solution:
    def partition(self, s: str) -> List[List[str]]:
        n = len(s)
        path = []
        res = []
        # lastIndx表示上一个回文子串的结束位置，当前就要从该位置开始往后找
        def dfs(lastIndx):
            if lastIndx == n-1:
                return [path.copy()]
            ans = []
            for i in range(lastIndx+1, n):
                tmp = s[lastIndx+1:i+1]
                if tmp == tmp[::-1]:
                    # 只有当该串回文时才继续往后找
                    path.append(tmp)
                    ans += dfs(i)
                    path.pop()
            return ans
        return dfs(-1)
        # def dfs(lastIndx):
        #     if lastIndx == n-1:
        #         res.append(path.copy())
        #         return
        #     for i in range(lastIndx+1, n):
        #         tmp = s[lastIndx+1:i+1]
        #         if tmp == tmp[::-1]:
        #             path.append(tmp)
        #             dfs(i)
        #             path.pop()
        # dfs(-1)
        # return res

if __name__ == "__main__":
    sl = Solution()
    s = "aab"
    print(sl.partition(s))
